### PR TITLE
Refactor `Tree.export`'s `on_node`

### PR DESCRIPTION
**Pull request by Ngoguey42**
On top of #1794

1. Refactors `on_node`

2. Addresses part of https://github.com/mirage/irmin/pull/1794#discussion_r830016058 by making sure that all the tree nodes exported go through `Node.export`.

3. Identifies 3 new edge cases which now result in exceptions. This led to modifications in `test/irmin/test_tree.ml | test_fold_force`.

### CODE REVIEW

File: src/irmin/tree.ml [hunk 1]
```diff
@@ -1986,32 +1984,6 @@ module Make (P : Backend.S) = struct
           false
       | Some false -> true
     in
-    let skip n =
-      match Node.cached_key n with
-      | Some k ->
-          cnt.node_mem <- cnt.node_mem + 1;
-          P.Node.mem node_t k
-      | None -> (
-          (* XXX: should we compute the actual hash here if it's not cached? *)
```
Comment (Ngoguey42) [inline src/irmin/tree.ml:1995]: No it should not be done because computing the hash requires exporting the children first.

**Comment by zshipko** [comment]
I just verified that this fixes the failing test in `irmin-server`

**Comment by icristescu** [comment]
I added a test that exposes the bug without relying on https://github.com/mirage/irmin/pull/1794, so the two PRs can be merged independently.

File: src/irmin/tree.ml [hunk 2]
```diff
@@ -2093,60 +2064,131 @@ module Make (P : Backend.S) = struct
...
+      | Pruned h ->
+          (* Case 3. [n] is a pruned hash. [P.Node.index node_t h] could be
+             different than [None], but let's always crash. *)
+          pruned_hash_exn "export" h
+      | Portable_dirty _ ->
+          (* Case 4. [n] is a portable value with diffs. The hash of the
+             reconstructed portable value could be known by [repo], but let's
+             always crash. *)
+          portable_value_exn "export"
+      | Map _ | Value _ | Key _ -> (
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin/tree.ml:2104]: Agree. 
If everything else looks good to you, I'll merge this today.
